src/wp-admin/includes/block-editor.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * Block Editor API.
+ *
+ * @package WordPress
+ * @subpackage Editor
+ * @since 5.8.0
+ */
+
+/**
+ * Gets the default block editor settings.
+ *
+ * @since 5.8.0
+ *
+ * @return array The default block editor settings.
+ */
+function get_default_block_editor_settings() {
+	/**
+	 * @global WP_Post $post Global post object.
+	 */
+	global $post;
+
+	/**
+	 * Filters the allowed block types for the editor, defaulting to true (all
+	 * block types supported).
+	 *
+	 * @since 5.0.0
+	 *
+	 * @param bool|array $allowed_block_types Array of block type slugs, or
+	 *                                        boolean to enable/disable all.
+	 * @param WP_Post    $post                The post resource data.
+	 */
+	$allowed_block_types = apply_filters( 'allowed_block_types', true, $post );
```

Comment:
We were discussing the `block_editor_settings` hook in this issue https://github.com/WordPress/gutenberg/issues/30082#issuecomment-803923854 with @nosolosw It seems the "context" argument proposed there can be useful for more than just the block_editor_settings hook, this one too maybe. (Might be too soon to introduce in this PR since there's no alternative block editor yet)

Reply:
Yes, we will need more granular controls. In part because some screens don't have any `$post` defined so there needs to be a different way to detect the context it deals with.

Reply:
Thinking a bit more about it, the biggest concern I have is that existing filters assume that `$post` contains an object, like here:
https://developer.wordpress.org/reference/hooks/allowed_block_types/#comment-3176
We even promote it ourselves in the documentation:
https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#hiding-blocks-from-the-inserter

We need to run this filter conditionally and provide an alternative for other screens.

Reply:
@SergeyBiryukov or @audrasjb, any hints on how we could make this hook more general so it works when `$post` is not defined? In general most of the hooks that filter things for blocks assume that `$post` is defined but it will no longer be true on the edit site, or edit widgets pages. If plugins use hooks that access properties of the post without checking if it exists then it will throw errors. Do we need to introduce new filters and map old versions?

Reply:
A good example of a possible approach is probably in the [filtering of rewrite rules](https://github.com/WordPress/wordpress-develop/blob/85907da8a4c8468046db7f42bc20fe7326f38dc3/src/wp-includes/class-wp-rewrite.php#L1421-L1433).

```
// new filter fires first
apply_filters( "allowed_block_types_{$context}", ...new args... );

if ( 'post' === $context ) {
    // old filter remains for the post context only, but deprecated.
    apply_filters_deprecated( 'allowed_block_types', array( true, $post ), '5.8.0', "allowed_block_types_post", __( 'Some message explaining' )  )
}
```

Reply:
Thank you, that's very helpful. Now, we only need to sort out what would the best approach to handle a new filter knowing that `$post` is optional depending on the context.

Reply:
I played with deprecations in https://github.com/WordPress/wordpress-develop/pull/1118/commits/8153420c73a766ef6c8ce91bb4f003f4fcf1ac9c.